syntax = "proto3";

package school;

option go_package = "genproto/payment";

import "google/protobuf/empty.proto";

service PaymentService {
    rpc CheckPerformTransaction(CheckPerformTransactionRequest) returns (CheckPerformTransactionResponse);
    rpc CreateTransaction(CreateTransactionRequest) returns (CreateTransactionResponse);
    rpc PerformTransaction(TransactionRequest) returns (PerformTransactionResponse);
    rpc CancelTransaction(CancelTransactionRequest) returns (CancelTransactionResponse);
    rpc CheckTransaction(TransactionRequest) returns (CheckTransactionResponse);
    rpc GetStatement(GetStatementRequest) returns (GetStatementResponse);
}

service ManagementService {
    rpc CreateBranch(CreateBranchRequest) returns (Branch);
    rpc GetBranch(ByIdRequest) returns (Branch);
    rpc ListBranches(google.protobuf.Empty) returns (ListBranchesResponse);
    rpc UpdateBranch(Branch) returns (Branch);
    rpc DeleteBranch(ByIdRequest) returns (google.protobuf.Empty);
    rpc CreateStudent(CreateStudentRequest) returns (Student);
    rpc GetStudentByAccountId(ByAccountIdRequest) returns (Student);
    rpc ListStudents(ListRequest) returns (ListStudentsResponse);
    rpc UpdateStudent(Student) returns (Student);
    rpc DeleteStudentByAccountId(ByAccountIdRequest) returns (google.protobuf.Empty);
    rpc CreateStudentsBatch(CreateStudentsBatchRequest) returns (CreateStudentsBatchResponse);
    rpc DeleteStudentsBatch(DeleteStudentsBatchRequest) returns (google.protobuf.Empty);
}

message Branch {
    string id = 1;
    string name = 2;
    int64 monthly_fee = 3;
    string mfo_code = 4;
    string account_number = 5;
    string merchant_id = 6;
}

message Student {
    string id = 1;
    string account_id = 2;
    string branch_id = 3;
    string parent_name = 4;
    double discount_percent = 5;
    int64 balance = 6;
    string full_name = 7;
    string group_name = 8;
    string phone = 9;
    string contract_number = 10;
    bool status = 11; // Talaba holati: true - o'qimoqda, false - o'qimaydi
}

message ByIdRequest { string id = 1; }
message ByAccountIdRequest { string account_id = 1; }
message ListRequest {}

message CreateBranchRequest {
    string name = 1;
    int64 monthly_fee = 2;
    string mfo_code = 3;
    string account_number = 4;
    string merchant_id = 5;
}

message ListBranchesResponse { repeated Branch branches = 1; }

message CreateStudentRequest {
    string account_id = 1;
    string branch_id = 2;
    string parent_name = 3;
    double discount_percent = 4;
    string full_name = 5;
    string group_name = 6;
    string phone = 7;
    string contract_number = 8;
}

message ListStudentsResponse { repeated Student students = 1; }

message CreateStudentsBatchRequest {
    repeated CreateStudentRequest students = 1;
}

message CreateStudentsBatchResponse {
    repeated Student students = 1; 
}

message DeleteStudentsBatchRequest {
    repeated string account_ids = 1;
}

message Account { string id = 1; }

message CheckPerformTransactionRequest {
    int64 amount = 1;
    Account account = 2;
}

message StudentAdditionalInfo {
    string full_name = 1;
    string parent_name = 2;
    string branch_name = 3;
    string group_name = 4;
    string group_number = 5;
    int64 amount_to_pay = 6;
    string mfo_code = 7;
    string account_number = 8;
}

message CheckPerformTransactionResponse {
    bool allow = 1;
    StudentAdditionalInfo additional = 2;
}

message CreateTransactionRequest {
    string id = 1;      // To'lov tizimidagi tranzaksiya IDsi
    int64 time = 2;     // Yaratilgan vaqt (timestamp)
    int64 amount = 3;   // Summa (tiyinda)
    Account account = 4; // Talaba hisobi
}

message Receiver {
    string id = 1;     // Merchant ID
    int64 amount = 2;  // Summa
}

message CreateTransactionResponse {
    int64 create_time = 1;
    string transaction = 2; // Tizimdagi ichki tranzaksiya IDsi
    int32 state = 3;
    repeated Receiver receivers = 4;
}

message TransactionRequest { string id = 1; }

message PerformTransactionResponse {
    string transaction = 1;
    int64 perform_time = 2;
    int32 state = 3;
}

message CancelTransactionRequest {
    string id = 1;
    int32 reason = 2;
}

message CancelTransactionResponse {
    string transaction = 1;
    int64 cancel_time = 2;
    int32 state = 3;
}

message CheckTransactionResponse {
    int64 create_time = 1;
    int64 perform_time = 2;
    int64 cancel_time = 3;
    string transaction = 4;
    int32 state = 5;
    optional int32 reason = 6;
}

message GetStatementRequest {
    int64 from = 1;
    int64 to = 2;
}

message StatementTransaction {
    string id = 1;
    int64 time = 2;
    int64 amount = 3;
    Account account = 4;
    int64 create_time = 5;
    int64 perform_time = 6;
    int64 cancel_time = 7;
    string transaction = 8;
    int32 state = 9;
    optional int32 reason = 10;
    repeated Receiver receivers = 11;
}

message GetStatementResponse {
    repeated StatementTransaction transactions = 1;
}